### ISSUE TITLE
Triggers in SQL

### ISSUE THREAD

**Issue by pgulutzan** [tarantool/tarantool#2356]
A problem with triggers in SQL.

box.sql.execute("CREATE TABLE b1 (s1 INT PRIMARY KEY, s2 INT);")
box.sql.execute("CREATE TABLE b2 (s1 INT PRIMARY KEY, s2 INT, FOREIGN KEY (s2)    REFERENCES b1 (s1));")
box.sql.execute("CREATE TABLE b3 (s1 VARCHAR(100) PRIMARY KEY);")
box.sql.execute("PRAGMA foreign_keys=on;")
box.sql.execute("CREATE TRIGGER b2_1 BEFORE INSERT ON b2 FOR EACH ROW WHEN   (SELECT COUNT(*) FROM b2) > 0 BEGIN INSERT INTO b3 VALUES ('before'); END;")
box.sql.execute("CREATE TRIGGER b2_2 AFTER INSERT ON b2 FOR EACH ROW WHEN (SELECT COUNT(*) FROM b2) > 0 BEGIN INSERT INTO b3 VALUES ('after'); END;")
box.sql.execute("INSERT INTO b2 VALUES (1,1);")
box.sql.execute("SELECT * FROM b3;")
box.sql.execute("SELECT * FROM b2;")

Result = the foreign-key check fails, there are two rows in table b3, and there is one row in table b2.
Actually there should be no rows in any tables, because the foreign-key
check failure should cause the INSERT to fail completely, but that is the
subject of a different issue. What matters here is: it seems that the
foreign-key check happens after the firing of the AFTER INSERT trigger.

**Comment by kyukhin** [comment]
Confirmed.

**Comment by kyukhin** [comment]
@pgulutzan, we cannot reproduce it anymore. Could you pls check if it is working for you?

**Comment by pgulutzan** [comment]
The example that I gave works correctly now. However, the comment that I made was: "What matters here is: it seems that the foreign-key check happens after the firing of the AFTER INSERT trigger." So try this example:
box.sql.execute([[CREATE TABLE b1 (s1 INT PRIMARY KEY, s2 INT UNIQUE);]])
box.sql.execute([[CREATE TABLE b2 (s1 INT PRIMARY KEY, s2 INT, FOREIGN KEY (s2) REFERENCES b1 (s1));]])
box.sql.execute([[PRAGMA foreign_keys=on;]])
box.sql.execute([[CREATE TRIGGER b2_2 AFTER INSERT ON b2 FOR EACH ROW BEGIN INSERT INTO b1 VALUES (1,1); END;]])
box.sql.execute([[INSERT INTO b2 VALUES (1,1);]])
Result: The insert into b2 succeeds.
Conclusion: FOREIGN KEY check is happening after AFTER trigger.
Therefore I believe this is not fixed.

**Comment by pgulutzan** [comment]
Now the FOREIGN KEY check happens at an earlier stage, so the original description of the "bug" is no longer applicable. I have corrected the descriptions of execution order in the SQL manual. I did not follow the description in https://github.com/tarantool/tarantool/wiki/SQL:-constraints-checks-order. If the description in the manual is now what will continue to be supported, then, although in my opinion the order is strange, this issue can be closed.

**Comment by kyukhin** [comment]
@pgulutzan, could you pls propose an order, which is correct from your point of view?

**Comment by pgulutzan** [comment]
Surely what's correct from my point of view is unimportant, but
I think you mean: why did I say that in my opinion the order
that is described in the Tarantool 1.8 SQL manual is strange?
I'll quote from it:
"
When INSERT or UPDATE or DELETE occurs for table X, Tarantool usually operates in this order:
For each row
  Perform constraint checks for any UNIQUE or PRIMARY KEY or CHECK clauses defined for table X.
  For each BEFORE trigger that refers to table X
	   Check that the trigger's WHEN condition is true.
	   Execute what is in the trigger's BEGIN|END block.
  Insert or update or delete the row in table X.
  For each AFTER trigger that refers to table X
	   Check that the trigger's WHEN condition is true.
	   Execute what is in the trigger's BEGIN|END block.
  Perform constraint checks for any FOREIGN KEY clauses defined for table X.
However, Tarantool does not guarantee execution order when there are multiple constraints,
 or multiple triggers for the same event (including NoSQL on_replace triggers
 or SQL INSTEAD OF triggers that affect a view of table X).
"
I have not checked today whether this is still true.
Everything is done within the main "For each row" loop;
this is unfortunate and non-standard but not a disaster.
The constraint checks, other than foreign-key constraint
checks, are happening so early that, if a foreign key or
an after trigger can cause a change, it can
cause a violation of the non-foreign-key constraints
-- that is what I think is strange.
I did say, though, that I suppose this can be closed.

**Comment by kyukhin** [comment]
This still looks like a bug to me.
Looks Oracle [1] FK checks are performed after AFTER trigger invocation.
In PostgreSQL [2] AFTER triggers are fired after all constraints are checked.

As I understand the Standard (4.39.2 Trigger execution):
```
For each state change SCj in SEC, the BEFORE triggers activated by SCj are executed before any of their triggering
events take effect. When those triggering events have taken effect, any AFTER triggers activated by the
state changes of SEC are executed.
```
AFTER triggers are fired after triggered event is taken effect.  Triggered events are
INSERT, UPDATE etc. So, AFTER triggers are fired after change takes place and
hence all constraints are checked. So, I guess PostgreSQL for rigid w.r.t. ANSI here.

@pgulutzan what do you think? Proposal is to defer AFTER triggers after all
constraints are checked.

Anyway, moving to the next MS.

[1] - https://www.postgresql.org/docs/9.1/static/sql-createtrigger.html
[2] - https://hourim.wordpress.com/2014/07/30/on-table-constraint-execution-order/

**Comment by Korablev77** [comment]
I suggest to delay this ticket or close it at all. Explanation: now it makes no sense to rework order of FK verification and triggers processing, since we are going to (re-)implement FK constraints and SQL triggers as native on_replace Tarantool triggers. So, fixing this issue now may lead to a lot of unnecessary code which will be removed/substituted later.
@kyukhin are you ok with my suggestion?

**Comment by kostja** [comment]
Isn't this issue a placeholder for re-work of triggers/constraings as native to tarantool storage?

**Comment by Korablev77** [comment]
No, it's not. Can't find issue, it is unlikely to be created yet. But this one is about order of executing FK constraints and SQL triggers, not about their re-implementation.

**Comment by kostja** [comment]
There can't be a ticket about re-implementing anything: most tickets have clear/measurable/user visible behavior change requests. This ticket is as good as any other to mandate moving trigger implementation into the core.

**Comment by Korablev77** [comment]
Ok, I won't argue. Could you then move this ticket to the next milestone?

**Comment by kyukhin** [comment]
> Ok, I won't argue. Could you then move this ticket to the next milestone?

Done.
